# 예외처리
- 예외 발생 시 프로그램의 비 정상 종료를 막고 정상적인 실행 상태를 유지하는 것
- 예외의 감지 및 예외 발생 시 동작할 코드 작성 필요함.

## 프로그램의 오류 ERROR
- 어떤 원인에 의해 오동작 하거나 비정상적으로 종료된 경우

### 발생 시점에 따른 구분
- 컴파일 에러: 컴파일 도중 문법의 오류러 컴파일러가 발견하는 오류
- 런타임 에러 : 컴파일은 성고하였으나 프로그램실행 중에 발생하는 오류
    - 문법에 오류는 없으나 비정상적인 종료
- 논리적 에러 : 컴파일도 되고 프로그램도 실행되나, 프로그램이 의도대로 동작하지 않는 경우. 프로그램이 정상적으로 실행되고 종료되었지만 결과가 예상과 다르게 동작하는 경우

## 에러와 예외
- 에러: 발생하면 복구할 수 없는 심각한 오류
    - 메모리 부족
    - 스택 오버 플로우

- 예외: 프로그래머가 적절한 코드를 통해 대비할 수 있는 오류
    - 클래스 형 변환 실패
    - 파일 읽기 실패

## 예외의 종류: 체크예외 / 언체크 예외
- 체크 예외: 컴파일 시점에서 예외 처리가 요구되는 예외
    - 컴파일 자체 x
- 언체크 예외: 생행 시점에서 발생하는 예외로, 컴파일러가 예외 처리를 강제하지 않음.
    - 프로그래머의 판단으로 처리하는것.


## 예외처리 키워드
- 직접 처리
    - try
    - catch
    - finally{}
- 간접처리
    - thorows
- 예외를 직접 발생시킬때
    - throw : 예외를 명시적 발생시키는 키워드

### 예외처리
- 프로그램 실생 시 발생할 수 있는 예외에 대한 대비코드 작성
- 프로그램의 비정상적인 종료를 막고 정상적인 상태를 유지할 수 있도록 함.
- 중괄호 생략 불가능
- 중첩된 예외처리: 블록 안에 또 다른 try~catch 구문이 올 수 있음

### 예외처리(try~catch구문)

```java
try{
    //1. 예외가 발생할 수 있는 코드
    // 2. 정상코드
} catch(SomeException e){
    // 3. 예외가 발생 했을 때 처리할 코드
}
// 4. 일반코드
```

- 예외발생시 :
    - catch 되면 :1 > 3 > 4
    - catch 되지 않으면 : 1 > 종료
- 예외 발생하지 않으면 : 1 > 2> 4


## 다중 예외처리시 유의 사항
- 예외 객체에 해당하는 catch 문장을 찾을 때는 다형성이 적용됨
- 상속 관계에서 상위 타입의 예외가 먼저 선언되는 경우 뒤에 등장하는 catch 블록은 동작할 기회가 없음
- 상속 관계가 없는 경우는 무관


## `메서드에서 예외 던지기`
- 메서드 선언부에 throws 키워드를 사용하여 예외 작성
- 예외가 여러 개일 경우 , 를 이용하여 나열
- 메서드가 예외를 처리하는 것이 아닌 전달
- 조상 타입의 예외러 처리 가능


```java
// 예외처리 예시 진행

public class ThrowsTest1 {
    // ChecedException: 컴파일러가 예외 처리를 강제!
    // try ~ catch로 예외처리르 직접 하던지, throws로 전달.
    // thorws의 의미?
    // 1. 미안하지만 이 메서드에서 처리하지 않을게. 
    // 이 메서드를 사용하는 사람이 예외 처리를 해서 쓰도록 해.
    // 2. 이 메서드를 사용한 프로그래머는 이 메서드는 checedexception을 발생시킬 수 있는 메서드구나
    // 내가 이 메서드를 쓰려면 내가 예외 처리 해야겠네
    // 3. checkedexception은 원래 무조건 처리해야되는데 메서드 선언에다가 thorws를 한번만 봐줄게임.
    // 그런데 이 메서드를 호출하는 곳에서는 예외에 관련된 내용을 처리해줘야해.
    // 프로그램의 시작점인 main 메서드 조차 throws하면 결국 예외는 처리되지 않음.
    // - 비정상적 종료가 되겠지?
    
    public static void main(String[] args) {
        try {
            method1();
        } catch (ClassNotFoundException e) {
            throw new RuntimeException(e);
        }

    }

    private static void method1() throws ClassNotFoundException {
        method2();
    }

    private static void method2() throws ClassNotFoundException {
        Class.forName("Hello");
    }
}
```

### Checked Exception / Unchecked Exception point
- Checked Exception
    - 필수 처리: checked exception을 throw할 경우, 해당 예외를 처리하거나 메서드 시그니처에 throws를 명시해야 한다. 이는 컴파일러가 체크하여 개발자가 예외 처리를 신경 쓰도록 유도하는것 이다.
- Unchecked Exception
    - 선택적 처리: unchecked exception은 throw할 수 있지만, 이 예외는 컴파일 타임에 체크되지 않으므로 개발자가 반드시 처리해야 하는 것은 아니다. 예를 들어, 프로그램의 버그로 인해 발생하는 NullPointerException을 throw할 수 있다.
#### 요약
- throw는 checked와 unchecked 모두에서 사용할 수 있지만, checked exception은 반드시 처리해야 하고, unchecked exception은 선택적으로 처리할 수 있다.

## 메서드 재정의(오버라이드)와 throws
- 메서드 재정의 시 조상클래스 메서드가 던지는 예외보다 부모 예외를 던질 수 없다.
    - 예를 들어, 부모 클래스의 메서드가 IOException을 던지면, 자식 클래스의 재정의된 메서드는 IOException이나 FileNotFoundException(하위 클래스 예외)은 던질 수 있지만, Exception(상위 클래스 예외)은 던질 수 없다.

## finally 구문
- finally는 예외 발생 여부와 상관 없이 언제나 실행됨.


## 자동 자원 반납 구문(try with resources)
- 자원 등을 반납할 때 finally에서 close()를 통해 반납을 주로 함.
- 코드가 지저분해지고 다른 예외 상황을 발생 시킬수도 있음.

```java
public class TryWithResources {
    public static void main(String[] args) {
        // try with resources 구문 사용
        try (FileInputStream fis = new FileInputStream("file.txt")) {
            // 파일을 읽는 로직 추가
            int data;
            while ((data = fis.read()) != -1) {
                System.out.print((char) data);
            }
        } catch (FileNotFoundException e) {
            System.out.println("파일을 찾을 수 없습니다.");
        } catch (IOException e) {
            System.out.println("입출력 오류 발생.");
        }
    }
}
```

- try() 괄호 안에 객체를 생성하는 코드를 작성하면 해당 객체는 close()를 호출하지 않아도 블록을 벗어난 순간 close()가 호출됨.
- 해당 객체의 클래스가 AutoCloseable이라는 인터페이스를 구현한 클래스이어야만 함

## 사용자 정의 예외
- 기존 정의된 예외 이외에 사용자가 직접 정의 예외를 작성할 수 있다.
- 대부분 Exception 또는 RuntimeException 클래스를 상속받아 작성할 수 있다.

 